crates/telio-nurse/src/aggregator.rs

```diff
@@ -0,0 +1,808 @@
+            },
+        )
+    }
+
+    #[allow(dead_code)]
+    pub fn change_peer_state_relayed(&mut self, event: AnalyticsEvent) {
+        self.change_peer_state_common(
+            event,
+            PeerEndpointTypes {
+                initiator_ep: EndpointType::Relay,
+                responder_ep: EndpointType::Relay,
+            },
+        )
+    }
+
+    fn change_peer_state_common(&mut self, event: AnalyticsEvent, endpoints: PeerEndpointTypes) {
+        if !self.aggregate_nat_traversal_events {
+            return;
+        }
+
+        let mut data_guard = self.data.write();
+        let event_entry = data_guard.current_peer_events.entry(event.public_key);
+        let new_segment = match event_entry.borrow() {
+            Entry::Occupied(entry)
+                if entry.get().0.peer_state != event.peer_state || entry.get().1 != endpoints =>
+            {
+                Some(ConnectionSegmentData::new_peer(
+                    &entry.get().0,
+                    entry.get().1,
+                    event.timestamp.into(),
+                    event.rx_bytes,
+                    event.tx_bytes,
+                ))
+            }
+            _ => None,
+        };
+
+        if event.peer_state != PeerState::Connected {
+            if let Entry::Occupied(entry) = event_entry {
+                entry.remove();
+            }
+        } else if let Entry::Vacant(entry) = event_entry {
+            entry.insert((event, endpoints));
+        }
+
+        data_guard.unacknowledged_segments.extend(new_segment);
+    }
+
+    #[allow(dead_code)]
+    pub fn change_relay_state(
+        &mut self,
+        event: AnalyticsEvent,
+        reason: RelayConnectionChangeReason,
+    ) {
+        if !self.aggregate_relay_events {
+            return;
+        }
+
+        let mut data_guard = self.data.write();
+
+        let new_segment = data_guard
+            .current_relay_event
+            .as_ref()
+            .filter(|old_event| {
+                old_event.public_key != event.public_key || old_event.peer_state != event.peer_state
+            })
+            .map(|old_event| ConnectionSegmentData::new_relay(old_event, event.timestamp.into()));
+
+        if let Some(segment) = new_segment {
+            data_guard.unacknowledged_segments.push(segment);
+        }
+
+        if new_segment.is_some() || data_guard.current_relay_event.is_none() {
+            data_guard.current_relay_event = Some(RelayEvent {
+                public_key: event.public_key,
+                peer_state: event.peer_state,
+                timestamp: event.timestamp.into(),
+                reason,
+            })
+        }
+    }
+
+    #[allow(dead_code)]
+    pub async fn collect_unacknowledged_segments(&self) -> Vec<ConnectionSegmentData> {
+        // Getting wg_interface is first, so it won't await while mutex is locked
+        let wg_interface = self.wg_interface.get_interface().await;
+        let mut data_guard = self.data.write();
```

Comment:
Non-async lock used in the async method can block tokio - please change the lock to the one from tokio.